common/channel/src/libra_channel.rs

```diff
@@ -0,0 +1,109 @@
+//! libra_channel provides an mpsc channel which has two ends `libra_channel::Receiver`
+//! and `libra_channel::Sender` similar to existing mpsc data structures.
+//! What makes it different from existing mpsc channels is that we have full control
+//! over how the internal queueing in the channel happens and how we schedule messages
+//! to be sent out from this channel.
+//! Note that libra_channel does not provide an implementation of this internal queueing
+//! mechanism, but provides a trait (`MessageQueue`) which the consumers of this channel
+//! can use to create libra_channels' according to their needs.
+
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::Waker;
+
+use futures::async_await::FusedStream;
+use futures::stream::Stream;
+use futures::task::Context;
+use futures::Poll;
+
+/// MessageQueue is a trait which provides a very simple set of methods for implementing
+/// a queue which will be used as the internal queue libra_channel.
+pub trait MessageQueue {
+    /// The actual type of the messages stored in this MessageQueue
+    type Message;
+
+    /// Push a message to this queue
+    fn push(&mut self, message: Self::Message);
+
+    /// Pop a message from this queue
+    fn pop(&mut self) -> Option<Self::Message>;
+}
+
+/// SharedState is a data structure private to this module which is
+/// shared by the sender and receiver.
+struct SharedState<T> {
+    /// The internal queue of messages in this Channel
+    internal_queue: T,
+
+    /// Waker is needed so that the Sender can notify the task executor/scheduler
+    /// that something has been pushed to the internal_queue and it ready for
+    /// consumption by the Receiver and then the executor/scheduler will wake up
+    /// the Receiver task to process the next item.
+    waker: Option<Waker>,
+}
+
+/// The sending end of the libra_channel.
+#[derive(Clone)]
+pub struct Sender<T> {
+    shared_state: Arc<Mutex<SharedState<T>>>,
+}
+
+impl<T: MessageQueue> Sender<T> {
+    /// This adds the message into the internal queue data structure. This is a non-blocking
+    /// synchronous call.
```

Comment:
"non-blocking" and "synchronous" don't quite make sense in the same sentence, especially here since this could block on the Mutex.

Reply:
Yea, we should remove the "non-blocking" part.

Reply:
Will do this in another PR